File: python/NuGramServer.py
#!/usr/bin/python
#
# NuGram Hosted Server client API in Python.
#
# Copyright (C) 2009 Nu Echo Inc.
#
# This code relies on the 'simplejson' library for compatibility
# between Python and Jython. It has been tested with Python 2.6 and
# Jython 2.5.0.
#
# simplejson can be found here:
#    http://code.google.com/p/simplejson
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:


import sys
import base64
import urllib
import httplib
from simplejson import JSONEncoder, JSONDecoder
from xml.dom import minidom, Node

## Some exception classes

class Error(Exception):
    pass

class InternalError(Error):
    pass

class AuthentificationError(Error):
    pass

class InstantiationError(Error):
    pass

class InterpretationError(Error):
    pass

class GetGrammarError(Error):
    pass


DEFAULT_GSERVER_HOST = "www.grammarserver.com"
DEFAULT_GSERVER_PORT = 8082


## An object of this class acts as a proxy to NuGram Hosted Server.

class GrammarServer:

    def __init__(self, host=DEFAULT_GSERVER_HOST, port=DEFAULT_GSERVER_PORT):
        self.host = host
        self.port = port

    def create_session(self, username, password):
        return Session(self, username, password)

    def get_url():
        return "http://" + self.host + ":" + self.port



## This class represents a session with NuGram Hosted Server.

class Session:

    def __init__(self, server, username, password):
        self.server = server
        self.username = username
        self.password = password
        self.sessionId = None
        self.initialize()


    def get_auth(self):
        return base64.b64encode(self.username + ':' + self.password)


    def request(self, url, mode, data={}, text=''):
        if text:
            data = text
        else:
            data = urllib.urlencode(data)

        connection = httplib.HTTPConnection(self.server.host, self.server.port)
        connection.request(mode, url, body=data, headers={'Authorization': 'Basic ' + self.get_auth()})
        response = connection.getresponse()
        return response.status, response.read()

    def initialize(self):
        status, content = self.request('/session', 'POST')

        if not (200 <= status < 300):
            if (400 <= status < 500) :
                raise AuthentificationError(str(status))
            if (500 <= status < 600) :
                raise InternalError(str(status))
            raise Error(str(status))

        xml = minidom.parseString(content)
        if xml:
            self.sessionId = xml.documentElement.getAttribute('id')
            return self.sessionId
        else:
            return None


    ## Returns the session ID
    def get_id():
        return self.sessionId


    ## This method uploads a source grammar to NuGram Hosted Server.
    def upload(self, name, content):
        url = '/grammar/' + name

        return self.request(url, 'PUT', text=content)


    ## This method requests NuGram Hosted Server to load a static grammar.
    def load(self, grammarPath):
        return self.instantiate(grammarPath)


    ## This method instantiates a dynamic grammar and loads it.
    def instantiate(self, grammarPath, context={}):
        url = '/grammar/' + self.sessionId + '/' + grammarPath

        jsonContext = JSONEncoder().encode(context)
        data = {'context': jsonContext, 'responseFormat': 'json'}
        status, content = self.request(url, 'POST', data)
     
        if not (200 <= status < 300):
            raise InstantiationError(str(status))
        
        response = eval(content)

        return InstantiatedGrammar(self, response['grammar'])

    def disconnect(self):
        url = '/session/' + self.sessionId
        return self.request(url, 'DELETE')


## Objects of this class act as proxy for instantiated grammars on NuGram Hosted Server.

class InstantiatedGrammar:

    # This constructor is called by a Session object. Should not
    # be called directly by client applications.
    def __init__(self, session, grammarInfo):
        self.session = session
        self.grammarInfo = grammarInfo

    
    ## Returns the URL of the grammar
    def get_url(extension='abnf'):
        url = self.session.server.get_url()
        if extension:
            url += "." + extension
        return url


    ## Retrieves the source representation of the grammar in the 
    ## requested format ('abnf', 'grxml', or 'gsl')
    def get_content(self, extension='abnf'):
        url = '/grammar:' + self.session.username + '/' + self.session.sessionId + '/' + self.grammarInfo['id']
        
        if extension: 
            url += '.' + extension
       
        status, content = self.session.request(url, 'GET')

        if not (200 <= status < 300):
            raise GetGrammarError(str(status))

        return content


    ## Computes the semantic interpretation of the given sentence
    ## (which must a string). Returns a Python object of 'False' if
    ## the sentence cannot be parsed by the grammar.
    def interpret(self, sentence):
        url = '/interpretation:' + self.session.username + '/' + self.session.sessionId + '/' + self.grammarInfo['id']

        data = {'sentence': sentence}
        data['responseFormat'] = 'json'

        status, content = self.session.request(url, 'POST', data)

        if not (200 <= status < 300):
            raise InterpretationError(str(status))

        response = JSONDecoder().decode(content)

        return response['interpretation']
 


## A complete example
##
## Call the following function with your username and password for NuGram Hosted Server

def test(username='user', password='passwd'):
    
    filename = 'testTest.abnf'
    grammar = '''\
#ABNF 1.0 UTF-8;

language en-US;
tag-format <semantics/1.0>;

root $digits;

public $digits  = 
@alt
    @for (digit : digits)
        @word digit
    @end
@end
;
'''

    server = GrammarServer()
    print "Opening a session with NuGram Hosted Server"
    session = server.create_session(username, password)
    print "Uploading a grammar.."
    session.upload(filename, grammar)

    print "Instantiating the dynamic grammar.."
    grammar = session.instantiate(filename,  {'digits':  ["one", "two", "three", "four"]} )
    otherGrammar = session.instantiate(filename, {'digits':  ["one", "two", "three", "four"]})

    print "Representation of the instantiated grammar in XML form  .."
    print grammar.get_content(extension='grxml')
    
    print "Interpreting the first grammar.."
    print grammar.interpret('one')

    print "Interpreting the second grammar.."
    print otherGrammar.interpret('two')
    print otherGrammar.interpret('no match')

    print "Terminating the session.."
    session.disconnect()


def usage():
    print "Usage: python NuGramServer.py username password"


if __name__ == '__main__':
    if len(sys.argv) == 3:
        test(*sys.argv[1:])
    else:
        usage()

File: javascript/nugram-server.js
/*
 * NuGram Hosted Server client API in JavaScript.
 *
 * Copyright (C) 2009 Nu Echo Inc.
 *
 * This code relies quite heavily on Rhino 1.7R2 as support
 * for ECMAScript for XML (E4X) is needed, as well as the
 * interface to Java classes.
 *
 * This implementation also relies on two other objects:
 *   * JSON - this object must provides two functions: parse and stringify,
 *            to convert JSON string to and from JavaScript objects. The
 *            accompanying file 'json.js' provides an implementation of this
 *            object.
 *
 *   * Base64 - Needed for HTTP Basic authorization. The accompanying file
 *            'base64.js' provides an implementation of this object. Not required
 *			  if the authorization token is given explicitly in argument
 *			  to the GrammarServer.createSession function (see documentation below). 
 * 
 * See at the end of this file for a test example. The most important functions
 * are documented below.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 */


/** 
    This function creates an object that acts as a proxy to NuGram Hosted Server.
    @param server the server name (defaults to 'www.grammarserver.com')
    @param port the server port (defaults to 8082)
**/
function GrammarServer(server, port, secure) {
    if (!server) server = 'www.grammarserver.com';
    if (!port) port = 8082;

    var protocol = secure ? "https://" : "http://";
    if (port == undefined) {
	this.url = protocol + server;
    }
    else {
        this.url = protocol + server + ":" + port;
    }
};

GrammarServer.prototype.getUrl = function () {
    return this.url;
};

/** 
    This function initiates a new session with NuGram Hosted Server and returns
    a GrammarServerSession object.
    
    If the password is omitted, the first argument must hold the
    authorization token for NuGram Hosted Server (the Base64 encoding
    of the string "username:password".

    @param username the account name on NuGram Hosted Server
    @param password the password for the account
**/
    
GrammarServer.prototype.createSession = function(username, password) {
	var authToken = username;
	if (password) {
    	var authToken = Base64.encode(username + ":" + password);
   	}
    return new GrammarServerSession(this, authToken);
};

function GrammarServerSession(server, authToken) {
    this.server = server;
    this.authToken = authToken;
    this.state = 'disconnected';
    this.sessionId = false;

    var result = __http_request(server.getUrl() + '/session', 'POST', authToken);
    if (result) {
        this.state = 'connected';
        this.sessionId = (new XML(result)).@id;
    }
};

GrammarServerSession.prototype.isConnected = function() {
    return this.state == 'connected';
};

GrammarServerSession.prototype.getSessionId = function() {
	return this.sessionId;
};

/**
   Uploads a grammar to NuGram Hosted Server.
   @param grammarPath the grammar path that will be used to refer to the grammar
   @param the text of the ABNF grammar

   Note: grammars are usually uploaded using NuGram IDE
   (see http://nugram.nuecho.com:8081/help/topic/com.nuecho.plugin.grammaride.doc/doc/html/abnf_publishing.html)
**/
GrammarServerSession.prototype.upload = function(grammarPath, content) {
    var result = __http_request(this.server.getUrl() + "/grammar/" + grammarPath, 'PUT', this.authToken, undefined, content);
    if (result) {
        return JSON.parse(result);
    }
    return false;
};


/**
   Loads a static grammar on NuGram Hosted Server. This is usually done when the application
   wants to do some semantic interpretation using the grammar. Returns an InstantiatedGrammar 
   object.

   @param grammarPath the path of the grammar to load
**/
GrammarServerSession.prototype.load = function(grammarPath) {
    return this.instantiate(grammarPath, {});
};


/**
   Instantiates a dynamic grammar with the given data. Returns an InstantiatedGrammar
   object upon successful instantiation.
   
   @param grammarPath the path of the grammar to instantiate
   @param the data used to populate the dynamic grammar template (the instantiation context)
**/
GrammarServerSession.prototype.instantiate = function (grammarPath, data) {
    if (this.sessionId) {
        var jsonData = JSON.stringify(data);
        var result = __http_request(this.server.getUrl() + "/grammar/" + this.sessionId + "/"  + grammarPath, 
                                    'POST', this.authToken, {context: jsonData, responseFormat: 'json'});
        if (result) {
            var jsonResult = JSON.parse(result);
            return new InstantiatedGrammar(this, jsonResult.grammar);
        }
        return false;
    }
    else {
        return false;
    }
};

/**
   Terminates the session with NuGram Hosted Server.
**/
GrammarServerSession.prototype.disconnect = function() {
    this.state = 'disconnected';
    __http_request(this.server.getUrl() + '/session/' + this.sessionId, 'DELETE', this.authToken);
};



function InstantiatedGrammar(session, data) {
    this.session = session;
    this.data = data;
};


/**
   Returns the URL of the instantiated grammar. This URL can be passed to an
   ASR engine.
   
   @param format (optional) the format of the source grammar. 
**/
InstantiatedGrammar.prototype.getUrl = function(format) {
    var url = this.data.grammarUrl;
    if (format) {
        url += '.';
        url += format;
    }
    return url;
}

/**
   Retrieves the source representation of the instantiated grammar.
   
   @param format (optional) the format of the source grammar to retrieve (defaults to ABNF)
**/
InstantiatedGrammar.prototype.getContent = function(format) {
    return __http_request(this.getUrl(format), 'GET');
};


/**
   Computes the semantic interpretation of the given sentence (a string) using the
   instantiated grammar. Returns an array of possible interpretations, or 'false' if
   there is none.
**/
InstantiatedGrammar.prototype.interpret = function (text) {
    var result = __http_request(this.data.interpreterUrl, 
                                'POST', 
                                this.session.authToken, 
                                {sentence: text.toString(), responseFormat: 'json'});
    if (result) {
        jsonResult = JSON.parse(result);
        return jsonResult.interpretation;
    }
    return false;
};


// *** A few helper functions

function __read_stream_as_string(inputStream) {
    var buffer = new java.lang.StringBuffer();
    var reader = new java.io.BufferedReader(new java.io.InputStreamReader(inputStream));

    while (reader.ready()) {
        buffer.append(reader.readLine()).append('\n');
    }
    return "" + buffer.toString();
};

function __http_request(url, method, authToken, data, text) {
    var connection = (new java.net.URL(url)).openConnection();
    connection.setRequestMethod(method);
    if (authToken) {
        connection.setRequestProperty("Authorization", "Basic " + authToken);
    }
    if (data || text) {
        var content = "";
        if (text) {
            content = text;
        }
        else {
            for (param in data) {
                content += "&" + param + "=" + data[param];
            }
            content = content.substring(1);
        }
        connection.setDoOutput(true);
        var writer = new java.io.OutputStreamWriter(connection.getOutputStream());
        writer.write(content);
        writer.close();
    }
    connection.setConnectTimeout(5000);
    if (connection.getResponseCode() >= 200 && connection.getResponseCode() < 300) {
        var response = __read_stream_as_string(connection.getInputStream());
        connection.disconnect();
        return response;
    }
    else {
        connection.disconnect();
        throw "com.nuecho.grammarserver.exception:" + connection.getResponseMessage();
    }
};




/**
   EXAMPLE
   
   A simple test function to illustrate the API usage.
   Simply call the test function with your username/password
   for NuGram Hosted Server.
**/

function test(username, password) {
    DIGITS_GRAMMAR = "#ABNF 1.0 ISO-8859-1;\n\
\n\
language en-US;\n\
tag-format <semantics/1.0>;\n\
\n\
root $digits;\n\
\n\
public $digits  = \n\
@alt\n\
    @for (digit : digits)\n\
        @word digit\n\
    @end\n\
@end\n\
;";

    gserver = new GrammarServer();

    print("Creating session....");
    session = gserver.createSession(username, password);
    print("Uploading grammar");
    session.upload("digits.abnf", DIGITS_GRAMMAR);

    print("Instantiating grammar");
    var grammar = session.instantiate("digits.abnf", {digits: ["one", "two", "three", "four"]});

    print("Grammar URL = " + grammar.getUrl());
    print("Grammar content:\n" + grammar.getContent('grxml'));

    var interp = grammar.interpret("two");
    print("Interpretation = " + interp);
    session.disconnect();
    return interp;
};

